spec/controllers/admin/customers_controller_spec.rb

```diff
@@ -171,4 +171,158 @@ def create_customer(enterprise)
+    let(:enterprise) { create(:distributor_enterprise) }
+    let(:customer) { create(:customer, enterprise: enterprise) }
+
+    before do
+      allow(controller).to receive(:spree_current_user) { enterprise.owner }
+    end
+
+    context 'when rendering json' do
+      context 'and the destroy succeeds' do
+        it 'invokes before destroy callbacks' do
+          allow(controller).to receive(:invoke_callbacks).with(:destroy, :after)
+
+          expect(controller).to receive(:invoke_callbacks).with(:destroy, :before)
+          spree_delete :destroy, id: customer, format: :json
+        end
+
+        it 'returns a no_content status code' do
+          spree_delete :destroy, id: customer, format: :json
+          expect(JSON.parse(response.body)['id']).to eq customer.id
+        end
+
+        it 'invokes after destroy callbacks' do
+          allow(controller).to receive(:invoke_callbacks).with(:destroy, :before)
+
+          expect(controller).to receive(:invoke_callbacks).with(:destroy, :after)
+          spree_delete :destroy, id: customer, format: :json
+        end
+      end
+
+      context 'and the destroy fails' do
+        before do
+          allow(Customer).to receive(:find).with(customer.id.to_s).and_return(customer)
+          allow(customer).to receive(:destroy).and_return(false)
+        end
+
+        it 'invokes before destroy callbacks' do
+          allow(controller).to receive(:invoke_callbacks).with(:destroy, :fails)
+
+          expect(controller).to receive(:invoke_callbacks).with(:destroy, :before)
+          spree_delete :destroy, id: customer, format: :json
+        end
+
+        it 'renders the errors' do
+          spree_delete :destroy, id: customer, format: :json
+          expect(response.body).to match('errors')
+        end
+
+        it 'returns conflict HTTP status code' do
+          spree_delete :destroy, id: customer, format: :json
+          expect(response).to have_http_status(:conflict)
+        end
+
+        it 'invokes failure destroy callbacks' do
+          allow(controller).to receive(:invoke_callbacks).with(:destroy, :before)
+
+          expect(controller).to receive(:invoke_callbacks).with(:destroy, :fails)
+          spree_delete :destroy, id: customer, format: :json
+        end
+      end
+    end
+
+    context 'when rendering html' do
+      context 'and the destroy succeeds' do
+        it 'invokes before destroy callbacks' do
+          allow(controller).to receive(:location_after_destroy) { '/fix_location_after_destroy' }
+          allow(controller).to receive(:invoke_callbacks).with(:destroy, :after)
+
+          expect(controller).to receive(:invoke_callbacks).with(:destroy, :before)
+          spree_delete :destroy, id: customer, format: :html
+        end
+
+        it 'raises due to missing #location_after_destroy' do
+          expect {
+            spree_delete :destroy, id: customer, format: :html
+          }.to raise_error(NameError, /location_after_destroy/)
```

Comment:
any how this method was implemented @mkllnk ? I couldn't find any trace including the metaprogramming I could think of.